File: agent-studio/server/routes/credits.js
/**
 * Credits API Routes
 * 
 * All credit-related API endpoints for prepaid billing system
 */

const express = require('express');
const { creditsService } = require('../services/creditsService');
const { paymentService } = require('../services/paymentService');

const router = express.Router();

// ========== CREDITS BALANCE ROUTES ==========

/**
 * GET /api/credits/balance
 * Get user credit balance
 */
router.get('/balance', async (req, res) => {
  try {
    const userId = req.user.uid;
    const balance = await creditsService.getUserBalance(userId);
    
    res.status(200).json({
      success: true,
      data: balance
    });
  } catch (error) {
    console.error('Error getting credit balance:', error);
    res.status(500).json({
      success: false,
      error: 'Failed to get credit balance'
    });
  }
});

/**
 * GET /api/credits/account
 * Get complete user credits account
 */
router.get('/account', async (req, res) => {
  try {
    const userId = req.user.uid;
    const account = await creditsService.getUserCreditsAccount(userId);
    
    res.status(200).json({
      success: true,
      data: account
    });
  } catch (error) {
    console.error('Error getting credits account:', error);
    res.status(500).json({
      success: false,
      error: 'Failed to get credits account'
    });
  }
});

// ========== CREDIT PACKAGES ROUTES ==========

/**
 * GET /api/credits/packages
 * Get available credit packages
 */
router.get('/packages', async (req, res) => {
  try {
    const packages = await creditsService.getCreditPackages();
    
    res.status(200).json({
      success: true,
      data: packages
    });
  } catch (error) {
    console.error('Error getting credit packages:', error);
    res.status(500).json({
      success: false,
      error: 'Failed to get credit packages'
    });
  }
});

/**
 * GET /api/credits/packages/:packageId
 * Get specific credit package
 */
router.get('/packages/:packageId', async (req, res) => {
  try {
    const packages = await creditsService.getCreditPackages();
    const package = packages.find(p => p.id === req.params.packageId);
    
    if (!package) {
      return res.status(404).json({
        success: false,
        error: 'Package not found'
      });
    }
    
    res.status(200).json({
      success: true,
      data: package
    });
  } catch (error) {
    console.error('Error getting credit package:', error);
    res.status(500).json({
      success: false,
      error: 'Failed to get credit package'
    });
  }
});

// ========== CREDIT PURCHASE ROUTES ==========

/**
 * POST /api/credits/purchase
 * Create credit purchase order
 */
router.post('/purchase', async (req, res) => {
  try {
    const userId = req.user.uid;
    const { packageId, quantity = 1, customAmount } = req.body;
    
    if (!packageId) {
      return res.status(400).json({
        success: false,
        error: 'Package ID is required'
      });
    }
    
    // Validate custom amount if provided
    if (packageId === 'custom_amount') {
      if (!customAmount || customAmount < creditsService.constants.MIN_PURCHASE) {
        return res.status(400).json({
          success: false,
          error: `Minimum purchase amount is ₹${creditsService.constants.MIN_PURCHASE}`
        });
      }
      
      if (customAmount > creditsService.constants.MAX_PURCHASE) {
        return res.status(400).json({
          success: false,
          error: `Maximum purchase amount is ₹${creditsService.constants.MAX_PURCHASE}`
        });
      }
    }
    
    // Create credit order
    const order = await creditsService.createCreditOrder(userId, packageId, quantity, customAmount);
    
    // Create payment with Razorpay
    const payment = await paymentService.createInvoicePayment(
      userId, 
      order.id, 
      order.total_amount
    );
    
    // Update order with payment details
    await creditsService.db.collection(creditsService.collections.CREDIT_ORDERS).doc(order.id).update({
      'payment.gateway_order_id': payment.id,
      'payment.payment_link': `https://rzp.io/i/${payment.id}`,
      updated_at: new Date().toISOString()
    });
    
    res.status(201).json({
      success: true,
      data: {
        order: {
          ...order,
          payment: {
            ...order.payment,
            gateway_order_id: payment.id,
            payment_link: `https://rzp.io/i/${payment.id}`
          }
        },
        payment_details: payment
      }
    });
  } catch (error) {
    console.error('Error creating credit purchase:', error);
    res.status(500).json({
      success: false,
      error: error.message || 'Failed to create credit purchase'
    });
  }
});

/**
 * POST /api/credits/purchase/verify
 * Verify payment and complete credit purchase
 */
router.post('/purchase/verify', async (req, res) => {
  try {
    const userId = req.user.uid;
    const { orderId, paymentId, signature } = req.body;
    
    if (!orderId || !paymentId || !signature) {
      return res.status(400).json({
        success: false,
        error: 'Order ID, payment ID, and signature are required'
      });
    }
    
    // Verify payment signature
    const isValidSignature = paymentService.verifyPaymentSignature(orderId, paymentId, signature);
    
    if (!isValidSignature) {
      return res.status(400).json({
        success: false,
        error: 'Invalid payment signature'
      });
    }
    
    // Complete the credit order
    const result = await creditsService.completeCreditOrder(orderId, {
      payment_id: paymentId,
      verified: true
    });
    
    res.status(200).json({
      success: true,
      data: result
    });
  } catch (error) {
    console.error('Error verifying credit purchase:', error);
    res.status(500).json({
      success: false,
      error: error.message || 'Failed to verify payment'
    });
  }
});

/**
 * GET /api/credits/orders
 * Get user's credit purchase orders
 */
router.get('/orders', async (req, res) => {
  try {
    const userId = req.user.uid;
    const limit = parseInt(req.query.limit) || 10;
    const status = req.query.status;
    
    let query = creditsService.db
      .collection(creditsService.collections.CREDIT_ORDERS)
      .where('userId', '==', userId)
      .orderBy('created_at', 'desc')
      .limit(limit);
    
    if (status) {
      query = query.where('status', '==', status);
    }
    
    const snapshot = await query.get();
    
    const orders = [];
    snapshot.forEach(doc => {
      orders.push({ id: doc.id, ...doc.data() });
    });
    
    res.status(200).json({
      success: true,
      data: orders
    });
  } catch (error) {
    console.error('Error getting credit orders:', error);
    res.status(500).json({
      success: false,
      error: 'Failed to get credit orders'
    });
  }
});

// ========== TRANSACTION HISTORY ROUTES ==========

/**
 * GET /api/credits/transactions
 * Get user transaction history
 */
router.get('/transactions', async (req, res) => {
  try {
    const userId = req.user.uid;
    const limit = parseInt(req.query.limit) || 20;
    const type = req.query.type; // purchase, usage, refund, bonus
    
    const transactions = await creditsService.getUserTransactions(userId, limit, type);
    
    res.status(200).json({
      success: true,
      data: transactions
    });
  } catch (error) {
    console.error('Error getting transactions:', error);
    res.status(500).json({
      success: false,
      error: 'Failed to get transactions'
    });
  }
});

/**
 * GET /api/credits/usage
 * Get usage summary for current period
 */
router.get('/usage', async (req, res) => {
  try {
    const userId = req.user.uid;
    const period = req.query.period || 'current_month';
    
    // Get usage transactions for the period
    const usageTransactions = await creditsService.getUserTransactions(userId, 100, 'usage');
    
    // Calculate usage summary
    let totalCreditsUsed = 0;
    let totalCharacters = 0;
    let totalApiCalls = 0;
    let agentUsage = {};
    
    const now = new Date();
    const startOfMonth = new Date(now.getFullYear(), now.getMonth(), 1);
    
    usageTransactions.forEach(transaction => {
      const transactionDate = new Date(transaction.timestamp);
      
      // Filter by period (for now, just current month)
      if (transactionDate >= startOfMonth) {
        totalCreditsUsed += Math.abs(transaction.amount);
        
        if (transaction.usage_details) {
          totalCharacters += transaction.usage_details.characters || 0;
          totalApiCalls += transaction.usage_details.api_calls || 0;
          
          const agentId = transaction.usage_details.agent_id;
          if (agentId) {
            if (!agentUsage[agentId]) {
              agentUsage[agentId] = {
                credits: 0,
                characters: 0,
                api_calls: 0
              };
            }
            agentUsage[agentId].credits += Math.abs(transaction.amount);
            agentUsage[agentId].characters += transaction.usage_details.characters || 0;
            agentUsage[agentId].api_calls += transaction.usage_details.api_calls || 0;
          }
        }
      }
    });
    
    res.status(200).json({
      success: true,
      data: {
        period,
        summary: {
          total_credits_used: totalCreditsUsed,
          total_characters: totalCharacters,
          total_api_calls: totalApiCalls,
          average_cost_per_character: totalCharacters > 0 ? totalCreditsUsed / totalCharacters : 0
        },
        agent_usage: agentUsage,
        period_start: startOfMonth.toISOString(),
        period_end: now.toISOString()
      }
    });
  } catch (error) {
    console.error('Error getting usage summary:', error);
    res.status(500).json({
      success: false,
      error: 'Failed to get usage summary'
    });
  }
});

// ========== CREDIT MANAGEMENT ROUTES ==========

/**
 * POST /api/credits/add
 * Add credits (admin only or promotional)
 */
router.post('/add', async (req, res) => {
  try {
    const { targetUserId, amount, bonusAmount = 0, reason } = req.body;
    const adminUserId = req.user.uid;
    
    if (!targetUserId || !amount || amount <= 0) {
      return res.status(400).json({
        success: false,
        error: 'Target user ID and valid amount are required'
      });
    }
    
    // TODO: Add admin permission check here
    
    const result = await creditsService.addCredits(
      targetUserId,
      amount,
      bonusAmount,
      {
        package_name: 'Admin Credit Addition',
        admin_action: true,
        admin_user_id: adminUserId,
        reason: reason || 'Manual credit addition'
      }
    );
    
    res.status(200).json({
      success: true,
      data: result
    });
  } catch (error) {
    console.error('Error adding credits:', error);
    res.status(500).json({
      success: false,
      error: 'Failed to add credits'
    });
  }
});

/**
 * GET /api/credits/status
 * Get comprehensive credit status for user
 */
router.get('/status', async (req, res) => {
  try {
    const userId = req.user.uid;
    
    const [account, lowBalanceCheck] = await Promise.all([
      creditsService.getUserCreditsAccount(userId),
      creditsService.checkLowBalanceAlert(userId)
    ]);
    
    res.status(200).json({
      success: true,
      data: {
        balance: account.balance,
        settings: account.settings,
        status: account.status,
        alerts: {
          low_balance: lowBalanceCheck.is_low_balance,
          threshold: lowBalanceCheck.threshold,
          should_alert: lowBalanceCheck.should_alert
        },
        can_use_services: account.balance?.total_credits > 0
      }
    });
  } catch (error) {
    console.error('Error getting credit status:', error);
    res.status(500).json({
      success: false,
      error: 'Failed to get credit status'
    });
  }
});

/**
 * PUT /api/credits/settings
 * Update credit account settings
 */
router.put('/settings', async (req, res) => {
  try {
    const userId = req.user.uid;
    const settings = req.body;
    
    // Update account settings
    await creditsService.db.collection(creditsService.collections.USER_CREDITS).doc(userId).update({
      settings: settings,
      updated_at: new Date().toISOString()
    });
    
    const updatedAccount = await creditsService.getUserCreditsAccount(userId);
    
    res.status(200).json({
      success: true,
      data: updatedAccount.settings
    });
  } catch (error) {
    console.error('Error updating credit settings:', error);
    res.status(500).json({
      success: false,
      error: 'Failed to update settings'
    });
  }
});

module.exports = router;

File: version-2.0/packages/infrastructure/billing-service/app/routes/credits.py
"""
Credits API routes
"""

import logging
from typing import List, Optional
from datetime import datetime, timezone

from fastapi import APIRouter, Depends, HTTPException, Request, Query
from fastapi.responses import JSONResponse
from google.cloud import firestore

from app.models.billing_models import (
    CreditPurchaseRequest, CreditAccount, CreditPackage, CreditOrder,
    CreditStatus, APIResponse, SettingsUpdateRequest
)
from app.services.firebase_service import FirebaseService
from app.services.credits_service import CreditsService
from app.services.payment_service import PaymentService
from app.middleware.auth_middleware import get_current_user

logger = logging.getLogger(__name__)
router = APIRouter()


def get_services(request: Request):
    """Get service dependencies"""
    firebase_service: FirebaseService = request.app.state.firebase
    credits_service = CreditsService(firebase_service)
    payment_service = PaymentService()
    return credits_service, payment_service


# ========== CREDITS BALANCE ROUTES ==========

@router.get("/balance")
async def get_credit_balance(
    request: Request,
    current_user: dict = Depends(get_current_user)
):
    """Get user credit balance"""
    try:
        credits_service, _ = get_services(request)
        user_id = current_user["uid"]
        
        balance = await credits_service.get_user_balance(user_id)
        
        return APIResponse(
            success=True,
            data=balance
        )
        
    except Exception as e:
        logger.error(f"Error getting credit balance: {e}")
        raise HTTPException(
            status_code=500,
            detail={
                "success": False,
                "error": "Failed to get credit balance",
                "details": str(e)
            }
        )


@router.get("/account")
async def get_credits_account(
    request: Request,
    current_user: dict = Depends(get_current_user)
):
    """Get complete user credits account"""
    try:
        credits_service, _ = get_services(request)
        user_id = current_user["uid"]
        
        account = await credits_service.get_user_credits_account(user_id)
        
        return APIResponse(
            success=True,
            data=account
        )
        
    except Exception as e:
        logger.error(f"Error getting credits account: {e}")
        raise HTTPException(
            status_code=500,
            detail={
                "success": False,
                "error": "Failed to get credits account",
                "details": str(e)
            }
        )


@router.get("/status")
async def get_credit_status(
    request: Request,
    current_user: dict = Depends(get_current_user)
):
    """Get comprehensive credit status for user"""
    try:
        credits_service, _ = get_services(request)
        user_id = current_user["uid"]
        
        # Get account and low balance check
        account = await credits_service.get_user_credits_account(user_id)
        low_balance_check = await credits_service.check_low_balance_alert(user_id)
        
        status_data = {
            "balance": account.get("balance", {}),
            "settings": account.get("settings", {}),
            "status": account.get("status", "active"),
            "alerts": low_balance_check,
            "can_use_services": account.get("balance", {}).get("total_credits", 0) > 0
        }
        
        return APIResponse(
            success=True,
            data=status_data
        )
        
    except Exception as e:
        logger.error(f"Error getting credit status: {e}")
        raise HTTPException(
            status_code=500,
            detail={
                "success": False,
                "error": "Failed to get credit status",
                "details": str(e)
            }
        )


# ========== CREDIT PACKAGES ROUTES ==========

@router.get("/packages")
async def get_credit_packages(request: Request):
    """Get available credit packages (public endpoint)"""
    try:
        credits_service, _ = get_services(request)
        
        packages = await credits_service.get_credit_packages()
        
        return APIResponse(
            success=True,
            data=packages
        )
        
    except Exception as e:
        logger.error(f"Error getting credit packages: {e}")
        raise HTTPException(
            status_code=500,
            detail={
                "success": False,
                "error": "Failed to get credit packages",
                "details": str(e)
            }
        )


@router.get("/packages/{package_id}")
async def get_credit_package(
    package_id: str,
    request: Request
):
    """Get specific credit package"""
    try:
        credits_service, _ = get_services(request)
        
        packages = await credits_service.get_credit_packages()
        package = next((p for p in packages if p["id"] == package_id), None)
        
        if not package:
            raise HTTPException(
                status_code=404,
                detail={
                    "success": False,
                    "error": "Package not found",
                    "error_code": "PACKAGE_NOT_FOUND"
                }
            )
        
        return APIResponse(
            success=True,
            data=package
        )
        
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error getting credit package: {e}")
        raise HTTPException(
            status_code=500,
            detail={
                "success": False,
                "error": "Failed to get credit package",
                "details": str(e)
            }
        )


# ========== CREDIT PURCHASE ROUTES ==========

@router.post("/purchase")
async def create_credit_purchase(
    purchase_request: CreditPurchaseRequest,
    request: Request,
    current_user: dict = Depends(get_current_user)
):
    """Create credit purchase order"""
    try:
        credits_service, payment_service = get_services(request)
        user_id = current_user["uid"]
        
        # Validate custom amount if needed
        if purchase_request.package_id == "custom_amount":
            if not purchase_request.custom_amount:
                raise HTTPException(
                    status_code=400,
                    detail={
                        "success": False,
                        "error": "Custom amount is required for custom packages",
                        "error_code": "MISSING_CUSTOM_AMOUNT"
                    }
                )
            
            if purchase_request.custom_amount < 1000:
                raise HTTPException(
                    status_code=400,
                    detail={
                        "success": False,
                        "error": f"Minimum purchase amount is ₹1000",
                        "error_code": "AMOUNT_TOO_LOW"
                    }
                )
            
            if purchase_request.custom_amount > 50000:
                raise HTTPException(
                    status_code=400,
                    detail={
                        "success": False,
                        "error": f"Maximum purchase amount is ₹50,000",
                        "error_code": "AMOUNT_TOO_HIGH"
                    }
                )
        
        # Create credit order
        order = await credits_service.create_credit_order(
            user_id,
            purchase_request.package_id,
            purchase_request.quantity,
            purchase_request.custom_amount
        )
        
        # Create payment with Razorpay
        payment = await payment_service.create_credit_purchase_order(
            user_id,
            order["id"],
            order["total_amount"],
            {
                "package_id": order["package_id"],
                "total_credits": order["total_credits"]
            }
        )
        
        # Update order with payment details
        await credits_service.firebase.update_credit_order(order["id"], {
            "payment.gateway_order_id": payment["id"],
            "payment.payment_link": f"https://rzp.io/i/{payment['id']}"
        })
        
        # Return updated order
        updated_order = {
            **order,
            "payment": {
                **order["payment"],
                "gateway_order_id": payment["id"],
                "payment_link": f"https://rzp.io/i/{payment['id']}"
            }
        }
        
        return APIResponse(
            success=True,
            data={
                "order": updated_order,
                "payment_details": payment
            }
        )
        
    except HTTPException:
        raise
    except ValueError as e:
        raise HTTPException(
            status_code=400,
            detail={
                "success": False,
                "error": str(e),
                "error_code": "VALIDATION_ERROR"
            }
        )
    except Exception as e:
        logger.error(f"Error creating credit purchase: {e}")
        raise HTTPException(
            status_code=500,
            detail={
                "success": False,
                "error": "Failed to create credit purchase",
                "details": str(e)
            }
        )


@router.get("/orders")
async def get_credit_orders(
    request: Request,
    current_user: dict = Depends(get_current_user),
    limit: int = Query(default=10, ge=1, le=100),
    status: Optional[str] = Query(default=None)
):
    """Get user's credit purchase orders"""
    try:
        credits_service, _ = get_services(request)
        user_id = current_user["uid"]
        
        # Get orders from Firebase
        firebase_service = credits_service.firebase
        
        from app.config.settings import settings
        query = firebase_service.db.collection(settings.CREDIT_ORDERS_COLLECTION)\
                                    .where("user_id", "==", user_id)\
                                    .limit(limit)
        
        if status:
            query = query.where("status", "==", status)
        
        docs = query.stream()
        
        orders = []
        for doc in docs:
            orders.append({"id": doc.id, **doc.to_dict()})
        
        return APIResponse(
            success=True,
            data=orders
        )
        
    except Exception as e:
        logger.error(f"Error getting credit orders: {e}")
        raise HTTPException(
            status_code=500,
            detail={
                "success": False,
                "error": "Failed to get credit orders",
                "details": str(e)
            }
        )


# ========== CREDIT MANAGEMENT ROUTES ==========

@router.post("/add")
async def add_credits(
    add_request: dict,
    request: Request,
    current_user: dict = Depends(get_current_user)
):
    """Add credits (admin only or promotional)"""
    try:
        credits_service, _ = get_services(request)
        
        target_user_id = add_request.get("target_user_id")
        base_amount = add_request.get("amount", 0)
        bonus_amount = add_request.get("bonus_amount", 0)
        reason = add_request.get("reason")
        
        if not target_user_id or base_amount <= 0:
            raise HTTPException(
                status_code=400,
                detail={
                    "success": False,
                    "error": "Target user ID and valid amount are required",
                    "error_code": "INVALID_REQUEST"
                }
            )
        
        # TODO: Add admin permission check here
        admin_user_id = current_user["uid"]
        
        result = await credits_service.add_credits(
            target_user_id,
            base_amount,
            bonus_amount,
            {
                "package_name": "Admin Credit Addition",
                "admin_action": True,
                "admin_user_id": admin_user_id,
                "reason": reason or "Manual credit addition"
            }
        )
        
        return APIResponse(
            success=True,
            data=result
        )
        
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error adding credits: {e}")
        raise HTTPException(
            status_code=500,
            detail={
                "success": False,
                "error": "Failed to add credits",
                "details": str(e)
            }
        )


@router.put("/settings")
async def update_credit_settings(
    settings_request: SettingsUpdateRequest,
    request: Request,
    current_user: dict = Depends(get_current_user)
):
    """Update credit account settings"""
    try:
        credits_service, _ = get_services(request)
        user_id = current_user["uid"]
        
        # Build settings update
        settings_update = {}
        if settings_request.low_balance_threshold is not None:
            settings_update["low_balance_threshold"] = settings_request.low_balance_threshold
        if settings_request.email_alerts is not None:
            settings_update["email_alerts"] = settings_request.email_alerts
        if settings_request.usage_alerts is not None:
            settings_update["usage_alerts"] = settings_request.usage_alerts
        
        # Update account settings
        await credits_service.firebase.db.collection(settings.USER_CREDITS_COLLECTION)\
                                          .document(user_id)\
                                          .update({
                                              "settings": settings_update,
                                              "updated_at": datetime.now(timezone.utc).isoformat()
                                          })
        
        # Get updated account
        updated_account = await credits_service.get_user_credits_account(user_id)
        
        return APIResponse(
            success=True,
            data=updated_account.get("settings", {})
        )
        
    except Exception as e:
        logger.error(f"Error updating credit settings: {e}")
        raise HTTPException(
            status_code=500,
            detail={
                "success": False,
                "error": "Failed to update settings",
                "details": str(e)
            }
        )


# ========== CREDIT USAGE ROUTES ==========

@router.post("/deduct")
async def deduct_credits(
    deduct_request: dict,
    request: Request,
    current_user: dict = Depends(get_current_user)
):
    """Deduct credits for usage (internal API)"""
    try:
        credits_service, _ = get_services(request)
        user_id = current_user["uid"]
        
        amount = deduct_request.get("amount", 0)
        usage_details = deduct_request.get("usage_details", {})
        
        if amount <= 0:
            raise HTTPException(
                status_code=400,
                detail={
                    "success": False,
                    "error": "Credit amount must be positive",
                    "error_code": "INVALID_AMOUNT"
                }
            )
        
        result = await credits_service.deduct_credits(user_id, amount, usage_details)
        
        return APIResponse(
            success=True,
            data=result
        )
        
    except ValueError as e:
        if "Insufficient credits" in str(e):
            raise HTTPException(
                status_code=402,
                detail={
                    "success": False,
                    "error": str(e),
                    "error_code": "INSUFFICIENT_CREDITS",
                    "actions": ["Purchase more credits", "Check current balance"]
                }
            )
        else:
            raise HTTPException(
                status_code=400,
                detail={
                    "success": False,
                    "error": str(e),
                    "error_code": "VALIDATION_ERROR"
                }
            )
    except Exception as e:
        logger.error(f"Error deducting credits: {e}")
        raise HTTPException(
            status_code=500,
            detail={
                "success": False,
                "error": "Failed to deduct credits",
                "details": str(e)
            }
        )